feat(general): cargo nightly, force repr c

bengreenier opened the pull request:
feat(repr-c): provides a flag for forcing `[repr(c)]` style layouts
   
This enables us to force `[repr(c)]` which can be useful (we'll use it internally) for forcing enum types to use C style layout, despite having a backing type. For us, the reason we can't easily just change to `[repr(c)]` is that it's in other generated code. We could consider updating that generator to be more specific, but for now this solves for what we need here.

feat(cargo): provides a flag for using `cargo +nightly` for expansion
    
This improves our dev experience where folks have `nightly` installed side by side, but use `stable` as default. Since we need expansion internally for some of our types to come through properly, this removes the need to switch toolchains before installing, instead letting cargo help us out there.

bengreenier committed:
feat(general): cargo nightly, force repr c

feat(repr-c): provides a flag for forcing `[repr(c)]` style layouts
feat(cargo): provides a flag for using `cargo +nightly` for expansion

bengreenier commented on Cargo.toml:
```diff
@@ -1,11 +1,12 @@
 [package]
 name = "cbindgen"
-version = "0.20.0"
+version = "0.20.0-rainway.0"
```
was thinking this might help us track what mainline version we're forked from, but definitely looking for feedback here.

mattiekat commented on src/bindgen/cargo/cargo.rs:
```diff
@@ -247,6 +250,19 @@ impl Cargo {
         )
     }
+
+    pub(crate) fn path() -> Result<String, std::io::Error> {
```
This should probably return a `Path` or a `PathBuf`. Did you need to replace the env var approach? That should have been correct.

bengreenier replied:
Yeah at least on Windows, it would appear that `CARGO` is `RUSTUP_HOME/{your_current_toolchain}/bin/cargo.exe` whereas this is `CARGO_HOME/bin/cargo.exe` - the former is tied to the toolchain and therefore doesn't support `+nightly` to easily use a sibling nightly toolchain.

```
# CARGO_HOME +nightly
λ C:\Users\bengr\.cargo\bin\cargo.exe +nightly --version
cargo 1.58.0-nightly (294967c53 2021-11-29)

# CARGO_HOME
λ C:\Users\bengr\.cargo\bin\cargo.exe --version
cargo 1.57.0 (b2e52d7ca 2021-10-21)

# RUSTUP_HOME +nightly
λ C:\Users\bengr\.rustup\toolchains\stable-x86_64-pc-windows-msvc\bin\cargo.exe +nightly --version
error: no such subcommand: `+nightly`

# RUSTUP_HOME
λ C:\Users\bengr\.rustup\toolchains\stable-x86_64-pc-windows-msvc\bin\cargo.exe --version
cargo 1.57.0 (b2e52d7ca 2021-10-21)
```


Went for string to avoid changing callers more - think it's worth changing to `PathBuf`?

bengreenier replied:
also just reverified after `rustup update` to be sure that version swap isn't related. confirmed it is not some weird `1.57` thing.

mattiekat commented on src/bindgen/config.rs:
```diff
@@ -594,6 +594,8 @@ pub struct EnumConfig {
     /// enums.
     pub private_default_tagged_enum_constructor: bool,
+    /// Whether to force all enums to use the C layout strategy, as if they have `[repr(C)]
+    pub force_repr_c: bool,
```
So just to be clear, in our case this is actually not the behavior we want. We want to force `repr(C)` for `repr(u32)` and `repr(i32)`, otherwise this will create FFI boundary issues when a Rust `repr(u8)` enum gets treated as an `int` in C.

bengreenier replied:
yeah, good point. This works for us today because the only odd-one out is our internal bebop structure for input levels (`repr(u32)`), but agree this has potential to break in other cases. I guess we have size here which should have this info, let me see if there's a quick fix.

bengreenier replied:
swapped this to a vec, taking primitives. e.g. `i32` - think that addresses this concern.

bengreenier committed:
feat(force_c): Accept a vec of primitives instead of bool

this allows the caller to more specifically configure which rust primitives to force c layouts for

bengreenier committed:
feat(conv): Support custom calling conventions

bengreenier committed:
Revert "feat(conv): Support custom calling conventions"

This reverts commit b4ddd02bb6f035ac0bac8ada0059a646b09b6161.

bengreenier committed:
feat(callconv): Add support for callconv

bengreenier commented:
second try looks good to me. that's what I get for staying up late to work on this.

bengreenier commented:
- [ ] last thing i'd like to add for us, drop `enum` in fn declaration if it's declared as `typedef enum MyEnum {} MyEnum;`

Update: Will punt on this for now and circle back, since it isn't strictly necessary.
